universe/db/postgres/rows.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package postgres
+
+import (
+	"context"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgproto3/v2"
+	"github.com/jackc/pgx/v4"
+)
+
+type hasQuery interface {
+	Query(context.Context, string, ...any) (pgx.Rows, error)
+}
+
+func query(ctx context.Context, opts commonOpts, q hasQuery, name, sql string, args ...any) (pgx.Rows, error) {
+	rows, err := returnWithSpan(ctx, opts, name, sql, func(ctx context.Context) (pgx.Rows, error) {
+		return q.Query(ctx, sql, args...)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return tracedRows{ctx, opts, sql, rows}, nil
+}
+
+type tracedRows struct {
+	ctx  context.Context
+	opts commonOpts
+	sql  string
+	rows pgx.Rows
+}
+
+func (r tracedRows) Close() {
+	r.rows.Close()
+}
+
+func (r tracedRows) Err() error {
+	return withSpan(r.ctx, r.opts, "db.Query.Err", r.sql, func(context.Context) error {
```

Comment:
It think it's more natural to put this into `Close()`.

`Err()` should be a pure read operation. `Close()` is the actual event.